File: iCyption2022/CftTest/transit/solution.py
n, m = map(int, input().split())

g = [[] for i in range(n)]

for i in range(m):
    a, b = map(int, input().split())
    g[a - 1].append(b - 1)

src, target = map(int, input().split())
src -= 1
target -= 1

q = []
vis = [False for i in range(n)]

q.append([src, 0])
vis[src] = True

while len(q) > 0:
    cur, dist = q[0]
    q.pop(0)

    if cur == target:
        print(dist - 1)
        quit(0)

    for adj in g[cur]:
        if not vis[adj]:
            q.append([adj, dist + 1])
            vis[adj] = True
print(-1)

File: iCyption2022/CftTest/transit/solution.cpp
#include <bits/stdc++.h>
using namespace std;
 
typedef long long ll;
typedef long double ld;
typedef pair<int,int> pii;
typedef pair<ll,ll> pll;
typedef vector<int> vi;
typedef vector<ll> vl;
typedef vector<vi> vvi;
typedef vector<vl> vvl;
typedef vector<bool> vb;
typedef vector<vb> vvb;
#define fi first
#define se second

void solve() {
    int n, m;
    cin >> n >> m;
    vvi g(n);
    for (int i = 0, x, y; i < m; i++) {
        cin >> x >> y, x--, y--;
        g[x].push_back(y);
    }
    int src, target;
    cin >> src >> target, src--, target--;

    queue<pii> q;
    bool vis[n] = {0};
    q.push({src, 0});
    vis[src] = true;
    while (!q.empty()) {
        auto [cur, dist] = q.front();
        q.pop();

        if (cur == target) {
            cout << dist - 1 << '\n';
            return;
        }

        for (int& adj : g[cur]) {
            if (!vis[adj]) {
                q.push({adj, dist + 1});
                vis[adj] = true;
            }
        }
    }
    cout << -1 << '\n';
}
 
int main() {
    ios::sync_with_stdio(0), cin.tie(0);
    solve();
}
